### ISSUE TITLE
AudioDeviceManager does not appropriately initialise AudioDeviceSetup.inputChannels for OboeAudioIODevice

### ISSUE THREAD

**Issue by atsushieno** [juce-framework/JUCE#837]
I'm porting JUCE audio plugin apps to Android, running as a standalone apps. I noticed that some of them crash, and figured out when - when they take audio inputs (almost effect plugins would) and when they enable Oboe by configuration.

I also figured out "why" - AudioDeviceManager does not appropriately initialise AudioDeviceSetup.inputChannels for OboeAudioIODevice. `AudioDeviceManager::initialise()` is passed `2` as numInputChannelsNeeded (correct), but at `AudioProcessorPlayer::audioDeviceAboutToStart()`, `numInputChans` is set as `1` (wrong), where the wrong value is from `(AudioIODevice*) device->getActiveInputChannels().countNumberOfSetBits()`. If Oboe is disabled, then it is AndroidAudioIODevice, and it correctly gives `2` there.

>  1) detailed steps on how to reproduce the bug, preferably with already existing JUCE code such as the demo plugin or the demo code

There is a reproducible CMake based audio plugin project at https://github.com/atsushieno/android-juce-cmake

The `main` branch does not enable Oboe and it works. There is `oboe-bug` branch which reproduces the bug, and is also easy to get the code diff: https://github.com/atsushieno/android-juce-cmake/compare/oboe-bug

>  2) the expected behaviour

It does not crash at initialising audio device

>  3) the operating system

Android (emulator, API Level 30, but can be anything)

>  4) the architecture (32-bit or 64-bit)

x64 (can ba anything)

>  5) a stack trace - if the bug causes a crash

```
juce::AudioProcessor::setPlayConfigDetails(int, int, double, int) juce_AudioProcessor.cpp:363
juce::AudioProcessorPlayer::setProcessor(juce::AudioProcessor*) juce_AudioProcessorPlayer.cpp:46
juce::StandalonePluginHolder::startPlaying() juce_StandaloneFilterWindow.h:229
juce::StandalonePluginHolder::init(bool, juce::String const&) juce_StandaloneFilterWindow.h:110
juce::StandalonePluginHolder::StandalonePluginHolder(juce::PropertySet*, bool, juce::String const&, juce::AudioDeviceManager::AudioDeviceSetup const*, juce::Array<juce::StandalonePluginHolder::PluginInOuts, juce::DummyCriticalSection, 0> const&, bool) juce_StandaloneFilterWindow.h:103
juce::StandaloneFilterWindow::StandaloneFilterWindow(juce::String const&, juce::Colour, juce::PropertySet*, bool, juce::String const&, juce::AudioDeviceManager::AudioDeviceSetup const*, juce::Array<juce::StandalonePluginHolder::PluginInOuts, juce::DummyCriticalSection, 0> const&, bool) juce_StandaloneFilterWindow.h:591
juce::StandaloneFilterApp::createWindow() juce_StandaloneFilterApp.cpp:81
juce::StandaloneFilterApp::initialise(juce::String const&) juce_StandaloneFilterApp.cpp:99
juce::JUCEApplicationBase::initialiseApp() juce_ApplicationBase.cpp:297
juce::JUCEApplication::initialiseApp() juce_Application.cpp:92
juce::JuceAppLifecycle::checkInitialised() juce_android_Messaging.cpp:284
juce::JuceAppLifecycle::onActivityResumed(_jobject*) juce_android_Messaging.cpp:246
juce::ActivityLifecycleCallbacks::invoke(_jobject*, _jobject*, _jobjectArray*) juce_android_JNIHelpers.cpp:432
juce::juce_invokeImplementer(_JNIEnv*, _jobject*, long long, _jobject*, _jobject*, _jobjectArray*) juce_android_JNIHelpers.cpp:401
art_quick_generic_jni_trampoline 0x00000000e3a46133
art_quick_invoke_stub 0x00000000e3a3f923
art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*) 0x00000000e3ad4382
art::interpreter::ArtInterpreterToCompiledCodeBridge(art::Thread*, art::ArtMethod*, art::ShadowFrame*, unsigned short, art::JValue*) 0x00000000e3c8a702
bool art::interpreter::DoCall<true, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7fcae
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a51569
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52b81
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::EnterInterpreterFromEntryPoint(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*) 0x00000000e3c7db01
artQuickToInterpreterBridge 0x00000000e408f326
art_quick_to_interpreter_bridge 0x00000000e3a4620e
art_quick_invoke_static_stub 0x00000000e3a3faf3
art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*) 0x00000000e3ad4393
art::JValue art::InvokeWithJValues<art::ArtMethod*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, art::ArtMethod*, jvalue const*) 0x00000000e3f32d84
art::JValue art::InvokeWithJValues<_jmethodID*>(art::ScopedObjectAccessAlreadyRunnable const&, _jobject*, _jmethodID*, jvalue const*) 0x00000000e3f330f6
art::JNI<true>::CallStaticObjectMethodA(_JNIEnv*, _jclass*, _jmethodID*, jvalue const*) 0x00000000e3dc580e
art::(anonymous namespace)::CheckJNI::CallMethodA(char const*, _JNIEnv*, _jobject*, _jclass*, _jmethodID*, jvalue const*, art::Primitive::Type, art::InvokeType) 0x00000000e3cfe097
art::(anonymous namespace)::CheckJNI::CallStaticObjectMethodA(_JNIEnv*, _jclass*, _jmethodID*, jvalue const*) 0x00000000e3ce950a
art::InvokeProxyInvocationHandler(art::ScopedObjectAccessAlreadyRunnable&, char const*, _jobject*, _jobject*, std::__1::vector<jvalue, std::__1::allocator<jvalue> >&) 0x00000000e4049a08
artQuickProxyInvokeHandler 0x00000000e409087b
art_quick_proxy_invoke_handler 0x00000000e3a45e47
art_quick_invoke_stub 0x00000000e3a3f923
art::ArtMethod::Invoke(art::Thread*, unsigned int*, unsigned int, art::JValue*, char const*) 0x00000000e3ad4382
art::interpreter::ArtInterpreterToCompiledCodeBridge(art::Thread*, art::ArtMethod*, art::ShadowFrame*, unsigned short, art::JValue*) 0x00000000e3c8a702
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea3f
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52b81
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52e4f
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a514de
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52e4f
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52e4f
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, true>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7f4b6
void art::interpreter::ExecuteSwitchImplCpp<true, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a8cd12
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73a8f
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, true>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7f4b6
void art::interpreter::ExecuteSwitchImplCpp<true, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a8cd12
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73a8f
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52e4f
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52e4f
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a514de
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52e4f
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52e4f
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, false>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7ea21
void art::interpreter::ExecuteSwitchImplCpp<false, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a52e4f
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73c82
art::interpreter::ArtInterpreterToInterpreterBridge(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*, art::JValue*) 0x00000000e3c7dc20
bool art::interpreter::DoCall<false, true>(art::ArtMethod*, art::Thread*, art::ShadowFrame&, art::Instruction const*, unsigned short, art::JValue*) 0x00000000e3c7f4b6
void art::interpreter::ExecuteSwitchImplCpp<true, false>(art::interpreter::SwitchImplContext*) 0x00000000e3a92614
ExecuteSwitchImplAsm 0x00000000e3a46de3
art::interpreter::Execute(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame&, art::JValue, bool, bool) (.llvm.16375758241455872412) 0x00000000e3c73a8f
art::interpreter::EnterInterpreterFromEntryPoint(art::Thread*, art::CodeItemDataAccessor const&, art::ShadowFrame*) 0x00000000e3c7db01
artQuickToInterpreterBridge 0x00000000e408f326
art_quick_to_interpreter_bridge 0x00000000e3a4620e
<unknown> 0x00000000e3a463b0
```

>  6) the plugin format (VST2, VST3, AU, AAX, RTAS) - if applicable

Standalone

>  7) which DAW you observed the bug in - if applicable

N/A

Confirmed with JUCE master/develop 383a423

**Comment by ed95** [comment]
This has been fixed in df06a471.

**Comment by atsushieno** [comment]
Awesome, thanks!

I assume issue #438 was also hitting the same issue (or, could be even an identical issue)
